# Regressão 01 - tarefa 03 - transformações em X e Y

Carregue os pacotes necessários e a base de gorjetas.

### I. Modelo no valor da gorjeta

1. Crie a matriz de design (e a matriz y) utilizando o Patsy, para um modelo em ```tip```, explicada por ```sex, smoker, diner e net_bill```.  
2. Remova as variáveis não significantes.  
3. observe o gráfico de resíduos em função de ```net_bill```  
4. teste transformar ```net_bill``` no log e um polinômio. Escolha o melhor modelo.

In [1]:
import pandas as pd
import seaborn as sns
from seaborn import load_dataset
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

import patsy
import statsmodels.api as sm
import statsmodels.formula.api as smf

%matplotlib inline

In [2]:
tips = sns.load_dataset("tips")
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']
tips.head()

total_bill   tip     sex smoker  day    time  size   tip_pct  net_bill
0       16.99  1.01  Female     No  Sun  Dinner     2  0.063204     15.98
1       10.34  1.66    Male     No  Sun  Dinner     3  0.191244      8.68
2       21.01  3.50    Male     No  Sun  Dinner     3  0.199886     17.51
3       23.68  3.31    Male     No  Sun  Dinner     2  0.162494     20.37
4       24.59  3.61  Female     No  Sun  Dinner     4  0.172069     20.98

In [3]:
# Importa a biblioteca Patsy e cria as matrizes de design para o modelo de regressão
# 'tip ~ sex + smoker + time + net_bill + 1' define a variável dependente e as independentes
# O operador '+ 1' inclui o intercepto no modelo
import patsy
y, X = patsy.dmatrices('tip ~ sex + smoker + time + net_bill + 1', tips)
X

DesignMatrix with shape (244, 5)
  Intercept  sex[T.Female]  smoker[T.No]  time[T.Dinner]  net_bill
          1              1             1               1     15.98
          1              0             1               1      8.68
          1              0             1               1     17.51
          1              0             1               1     20.37
          1              1             1               1     20.98
          1              0             1               1     20.58
          1              0             1               1      6.77
          1              0             1               1     23.76
          1              0             1               1     13.08
          1              0             1               1     11.55
          1              0             1               1      8.56
          1              1             1               1     30.26
          1              0             1               1     13.85
          1              0   

In [4]:
# Ajusta um modelo de Regressão Linear (OLS) com as matrizes de design criadas anteriormente
# 'y' é a variável dependente e 'X' são as variáveis independentes
modelo = sm.OLS(y,X).fit()

modelo.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    tip   R-squared:                       0.335
Model:                            OLS   Adj. R-squared:                  0.324
Method:                 Least Squares   F-statistic:                     30.13
Date:                Wed, 02 Oct 2024   Prob (F-statistic):           2.64e-20
Time:                        09:13:26   Log-Likelihood:                -375.13
No. Observations:                 244   AIC:                             760.3
Df Residuals:                     239   BIC:                             777.8
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          1.1999      0.242      4.952      0.000       0.723       1.677
sex[T.Female]     -0.0034      0.156     -0.022      0.982      -0.311       0.305
smoker[T.No]       0.1405      0.151      0.932      0.352      -0.156       0.437
time[T.Dinner]     0.0584      0.168      0.347      0.729      -0.273       0.390
net_bill           0.0995      0.009     10.662      0.000       0.081       0.118
==============================================================================
Omnibus:                       36.537   Durbin-Watson:                   2.121
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               65.875
Skew:                           0.800   Prob(JB):                     4.96e-15
Kurtosis:                       4.980   Cond. No.                         71.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### 1.R-quadrado e R-quadrado ajustado:


* **R-quadrado**: 0,335: O modelo explica cerca de 33,5% da variabilidade da variável dependente (tip). Isso significa que 33,5% da variação nas gorjetas pode ser explicada pelas variáveis independentes incluídas no modelo.


* **R-quadrado** ajustado: 0,324: Este valor ajusta o R-quadrado para o número de variáveis no modelo e o tamanho da amostra. A proximidade do valor ajustado com o R-quadrado simples sugere que o modelo não está superajustado.

#### 2. Estatística F e Probabilidade F:


* **F-statistic**: 30.13 e Prob (F-statistic): 2.64e-20: Esses valores indicam que o modelo como um todo é altamente significativo. A probabilidade associada à estatística F é praticamente zero, o que significa que pelo menos uma das variáveis independentes tem um efeito significativo na variável dependente.


#### 3. Coeficientes (coef), Erros padrão (std err) e p-values (P>|t|):


* **Intercepto**: O valor do intercepto é 1,1999, o que significa que, em média, a gorjeta esperada quando todas as variáveis independentes são zero é aproximadamente 1,20.


* **sex[T.Female]**: O coeficiente de -0,0034 sugere que o sexo não tem um efeito significativo nas gorjetas, com um p-value de 0,982, muito maior que 0,05.


* **smoker[T.No]**: O coeficiente de 0,1405 sugere que não fumar pode aumentar as gorjetas em média, mas o p-value de 0,352 indica que esse efeito não é estatisticamente significativo.


* **time[T.Dinner]**: O coeficiente de 0,0584 sugere que o jantar pode aumentar a gorjeta, mas com um p-value de 0,729, essa variável também não é significativa.


* **net_bill**: O coeficiente de 0,0995 é altamente significativo (p-value de 0,000), indicando que o valor da conta líquida (sem gorjeta) tem um impacto forte e positivo nas gorjetas. Para cada aumento de 1 unidade no valor da conta, a gorjeta aumenta em 0,0995.


#### 4. Estatísticas de diagnóstico:

* **Omnibus e Prob(Omnibus)**: A estatística de Omnibus testa a normalidade dos resíduos. Com um valor de 36,537 e um p-value de 0,000, há evidência de que os resíduos não seguem uma distribuição normal, o que pode indicar problemas com a suposição de normalidade do erro.


* **Durbin-Watson**: 2,121: Isso testa a presença de autocorrelação nos resíduos. Um valor próximo de 2, como aqui, sugere que não há autocorrelação significativa.


* **Jarque-Bera (JB) e Prob(JB)**: Testa a normalidade dos resíduos. Com um valor de 65,875 e p-value de 4.96e-15, também sugere que os resíduos não são normalmente distribuídos.


* **Skew**: 0,800 e Kurtosis: 4,980: O valor de skew (assimetria) e kurtosis (curtose) indicam que os resíduos são ligeiramente assimétricos e com caudas mais pesadas que a distribuição normal.


#### 5. Cond. No. (Número de Condicionamento):


* **Cond. No. 71.2**: Este valor avalia a multicolinearidade entre as variáveis. Um valor muito alto (geralmente > 30) pode indicar problemas de multicolinearidade, mas 71,2 é um valor moderado, sugerindo que a multicolinearidade não é um grande problema no modelo.

#### Conclusão:


* O modelo é globalmente significativo, conforme indicado pela estatística F.


* A única variável estatisticamente significativa é net_bill, o que indica que o valor da conta líquida é um bom preditor da gorjeta.


* As variáveis sexo, fumante e hora do dia (almoço ou jantar) não são estatisticamente 
significativas, o que sugere que não influenciam significativamente as gorjetas neste modelo.


* Há alguns sinais de que os resíduos não seguem perfeitamente a suposição de normalidade (Omnibus, Jarque-Bera), mas isso pode ser aceitável dependendo do contexto da análise.

In [5]:
# criando a martiz desing apenas com a variável de interesse
y, X = patsy.dmatrices('tip ~  net_bill + 1', tips)
X

DesignMatrix with shape (244, 2)
  Intercept  net_bill
          1     15.98
          1      8.68
          1     17.51
          1     20.37
          1     20.98
          1     20.58
          1      6.77
          1     23.76
          1     13.08
          1     11.55
          1      8.56
          1     30.26
          1     13.85
          1     15.43
          1     11.81
          1     17.66
          1      8.66
          1     12.58
          1     13.47
          1     17.30
          1     13.84
          1     17.54
          1     13.54
          1     31.84
          1     16.64
          1     15.47
          1     11.37
          1     10.69
          1     17.40
          1     16.65
  [214 rows omitted]
  Terms:
    'Intercept' (column 0)
    'net_bill' (column 1)
  (to view full data, use np.asarray(this_obj))

In [6]:
modelo = sm.OLS(y,X).fit()

modelo.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    tip   R-squared:                       0.333
Model:                            OLS   Adj. R-squared:                  0.330
Method:                 Least Squares   F-statistic:                     120.6
Date:                Wed, 02 Oct 2024   Prob (F-statistic):           5.02e-23
Time:                        09:13:26   Log-Likelihood:                -375.63
No. Observations:                 244   AIC:                             755.3
Df Residuals:                     242   BIC:                             762.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.3307      0.168      7.907      0.000       0.999       1.662
net_bill       0.0993      0.009     10.980      0.000       0.082       0.117
==============================================================================
Omnibus:                       33.873   Durbin-Watson:                   2.123
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               59.942
Skew:                           0.754   Prob(JB):                     9.63e-14
Kurtosis:                       4.903   Cond. No.                         43.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [7]:
# criando a martiz desing apenas com a variável de interesse e transformando para np_log
y, X = patsy.dmatrices('tip ~  np.log(net_bill) + 1', tips)
X

DesignMatrix with shape (244, 2)
  Intercept  np.log(net_bill)
          1           2.77134
          1           2.16102
          1           2.86277
          1           3.01406
          1           3.04357
          1           3.02432
          1           1.91250
          1           3.16800
          1           2.57108
          1           2.44669
          1           2.14710
          1           3.40983
          1           2.62829
          1           2.73631
          1           2.46895
          1           2.87130
          1           2.15871
          1           2.53211
          1           2.60046
          1           2.85071
          1           2.62756
          1           2.86448
          1           2.60565
          1           3.46072
          1           2.81181
          1           2.73890
          1           2.43098
          1           2.36931
          1           2.85647
          1           2.81241
  [214 rows omitted]
  Terms:
    'In

In [8]:
modelo = sm.OLS(y,X).fit()

modelo.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    tip   R-squared:                       0.294
Model:                            OLS   Adj. R-squared:                  0.292
Method:                 Least Squares   F-statistic:                     101.0
Date:                Wed, 02 Oct 2024   Prob (F-statistic):           4.40e-20
Time:                        09:13:26   Log-Likelihood:                -382.40
No. Observations:                 244   AIC:                             768.8
Df Residuals:                     242   BIC:                             775.8
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept           -1.2575      0.430     -2.924      0.004      -2.104      -0.410
np.log(net_bill)     1.5695      0.156     10.049      0.000       1.262       1.877
==============================================================================
Omnibus:                       73.945   Durbin-Watson:                   2.045
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              221.148
Skew:                           1.303   Prob(JB):                     9.51e-49
Kurtosis:                       6.868   Cond. No.                         17.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### II. Modelo no valor do percentual da gorjeta

1. Crie a matriz de design (e a matriz y) utilizando o Patsy, para um modelo no log de ```tip```, explicado por ```sex, smoker, diner e net_bill```.
2. Remova as variáveis não significantes.
3. Observe o gráfico de resíduos em função de ```net_bill```
4. Teste transformar ```net_bill``` no log e um polinômio. Escolha o melhor modelo.
5. Do modelo final deste item, calcule o $R^2$ na escala de ```tip``` (sem o log). Compare com o modelo do item 1. Qual tem melhor coeficiente de determinação?

In [9]:
tips = sns.load_dataset("tips")
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']
tips.head()

total_bill   tip     sex smoker  day    time  size   tip_pct  net_bill
0       16.99  1.01  Female     No  Sun  Dinner     2  0.063204     15.98
1       10.34  1.66    Male     No  Sun  Dinner     3  0.191244      8.68
2       21.01  3.50    Male     No  Sun  Dinner     3  0.199886     17.51
3       23.68  3.31    Male     No  Sun  Dinner     2  0.162494     20.37
4       24.59  3.61  Female     No  Sun  Dinner     4  0.172069     20.98

In [10]:
# Importa a biblioteca Patsy e cria as matrizes de design para o modelo de regressão
# 'tip ~ sex + smoker + time + net_bill + 1' define a variável dependente e as independentes
# O operador '+ 1' inclui o intercepto no modelo
y, X = patsy.dmatrices('tip_pct ~ sex + smoker + time + net_bill + 1', tips)
X

DesignMatrix with shape (244, 5)
  Intercept  sex[T.Female]  smoker[T.No]  time[T.Dinner]  net_bill
          1              1             1               1     15.98
          1              0             1               1      8.68
          1              0             1               1     17.51
          1              0             1               1     20.37
          1              1             1               1     20.98
          1              0             1               1     20.58
          1              0             1               1      6.77
          1              0             1               1     23.76
          1              0             1               1     13.08
          1              0             1               1     11.55
          1              0             1               1      8.56
          1              1             1               1     30.26
          1              0             1               1     13.85
          1              0   

In [11]:
# Ajusta um modelo de Regressão Linear (OLS) com as matrizes de design criadas anteriormente
# 'y' é a variável dependente e 'X' são as variáveis independentes
modelo = sm.OLS(y,X).fit()

modelo.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.107
Model:                            OLS   Adj. R-squared:                  0.092
Method:                 Least Squares   F-statistic:                     7.157
Date:                Wed, 02 Oct 2024   Prob (F-statistic):           1.85e-05
Time:                        09:13:26   Log-Likelihood:                 110.12
No. Observations:                 244   AIC:                            -210.2
Df Residuals:                     239   BIC:                            -192.8
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          0.3216      0.033      9.698      0.000       0.256       0.387
sex[T.Female]     -0.0069      0.021     -0.321      0.749      -0.049       0.035
smoker[T.No]      -0.0351      0.021     -1.700      0.090      -0.076       0.006
time[T.Dinner]     0.0221      0.023      0.961      0.337      -0.023       0.067
net_bill          -0.0066      0.001     -5.191      0.000      -0.009      -0.004
==============================================================================
Omnibus:                      461.972   Durbin-Watson:                   2.032
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           218184.236
Skew:                          10.789   Prob(JB):                         0.00
Kurtosis:                     147.897   Cond. No.                         71.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [12]:
y, X = patsy.dmatrices('tip_pct ~ np.log(net_bill) + 1', tips)

modelo = sm.OLS(y,X).fit()

modelo.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.202
Model:                            OLS   Adj. R-squared:                  0.198
Method:                 Least Squares   F-statistic:                     61.16
Date:                Wed, 02 Oct 2024   Prob (F-statistic):           1.62e-13
Time:                        09:13:26   Log-Likelihood:                 123.81
No. Observations:                 244   AIC:                            -243.6
Df Residuals:                     242   BIC:                            -236.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept            0.6181      0.054     11.444      0.000       0.512       0.724
np.log(net_bill)    -0.1534      0.020     -7.820      0.000      -0.192      -0.115
==============================================================================
Omnibus:                      440.899   Durbin-Watson:                   1.965
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           171509.052
Skew:                           9.850   Prob(JB):                         0.00
Kurtosis:                     131.381   Cond. No.                         17.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [18]:
# Substitua o log por termos polinomiais de ordem 2
y, X = patsy.dmatrices('tip_pct ~ net_bill + I(net_bill ** 2)', tips)

# Ajuste o modelo OLS
modelo = sm.OLS(y, X).fit()

modelo.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.158
Model:                            OLS   Adj. R-squared:                  0.151
Method:                 Least Squares   F-statistic:                     22.55
Date:                Wed, 02 Oct 2024   Prob (F-statistic):           1.06e-09
Time:                        09:20:38   Log-Likelihood:                 117.25
No. Observations:                 244   AIC:                            -228.5
Df Residuals:                     241   BIC:                            -218.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept            0.4851      0.047     10.332      0.000       0.393       0.578
net_bill            -0.0268      0.005     -5.477      0.000      -0.036      -0.017
I(net_bill ** 2)     0.0005      0.000      4.356      0.000       0.000       0.001
==============================================================================
Omnibus:                      447.620   Durbin-Watson:                   2.026
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           186035.672
Skew:                          10.140   Prob(JB):                         0.00
Kurtosis:                     136.744   Cond. No.                     2.42e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.42e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [20]:
from sklearn.metrics import r2_score
# Prever os valores com base no modelo ajustado
tips['pred_tip_pct'] = modelo.predict(X)

# Calcular o valor predito de 'tip' com base no 'tip_pct' previsto
tips['pred_tip'] = tips['pred_tip_pct'] * tips['total_bill']

# Calcular o R² na escala original de 'tip'
r2_original = r2_score(tips['tip'], tips['pred_tip'])

print(f"R² na escala de 'tip' (sem log): {r2_original}")

R² na escala de 'tip' (sem log): 0.05628784530547193


### III. Previsão de renda

Vamos trabalhar a base que você vai usar no projeto do final deste ciclo.

Carregue a base ```previsao_de_renda.csv```.

|variavel|descrição|
|-|-|
|data_ref                | Data de referência de coleta das variáveis |
|index                   | Código de identificação do cliente|
|sexo                    | Sexo do cliente|
|posse_de_veiculo        | Indica se o cliente possui veículo|
|posse_de_imovel         | Indica se o cliente possui imóvel|
|qtd_filhos              | Quantidade de filhos do cliente|
|tipo_renda              | Tipo de renda do cliente|
|educacao                | Grau de instrução do cliente|
|estado_civil            | Estado civil do cliente|
|tipo_residencia         | Tipo de residência do cliente (própria, alugada etc)|
|idade                   | Idade do cliente|
|tempo_emprego           | Tempo no emprego atual|
|qt_pessoas_residencia   | Quantidade de pessoas que moram na residência|
|renda                   | Renda em reais|

1. Ajuste um modelo de regressão linear simples para explicar ```renda``` como variável resposta, por ```tempo_emprego``` como variável explicativa. Observe que há muitas observações nessa tabela. Utilize os recursos que achar necessário.
2. Faça uma análise de resíduos. Com os recursos vistos neste módulo, como você melhoraria esta regressão?
3. Ajuste um modelo de regressão linear múltipla para explicar ```renda``` (ou uma transformação de ```renda```) de acordo com as demais variáveis.
4. Remova as variáveis não significantes e ajuste novamente o modelo. Interprete os parâmetros
5. Faça uma análise de resíduos. Avalie a qualidade do ajuste.

In [21]:
df = pd.read_csv('previsao_de_renda.csv')
df.head()

Unnamed: 0    data_ref  id_cliente sexo  posse_de_veiculo  posse_de_imovel  \
0           0  2015-01-01       15056    F             False             True   
1           1  2015-01-01        9968    M              True             True   
2           2  2015-01-01        4312    F              True             True   
3           3  2015-01-01       10639    F             False             True   
4           4  2015-01-01        7064    M              True            False   

   qtd_filhos        tipo_renda           educacao estado_civil  \
0           0        Empresário         Secundário     Solteiro   
1           0       Assalariado  Superior completo       Casado   
2           0        Empresário  Superior completo       Casado   
3           1  Servidor público  Superior completo       Casado   
4           0       Assalariado         Secundário     Solteiro   

  tipo_residencia  idade  tempo_emprego  qt_pessoas_residencia    renda  
0            Casa     26       6.602740                    1.0  8060.34  
1            Casa     28       7.183562                    2.0  1852.15  
2            Casa     35       0.838356                    2.0  2253.89  
3            Casa     30       4.846575                    3.0  6600.77  
4   Governamental     33       4.293151                    1.0  6475.97